### PR TITLE
Bring emit-h back to life!

### PR DESCRIPTION
**Pull request by SuperAuguste**
See the current results here: https://zigbin.io/a90812

Some things I still need to do:
- [x] Use export names rather than the names of the exported decls (can be different w/ `@export` for example)
- [x] ~~Give structs better generated names(?)~~ Decided to keep simple, not fully-qualified names for now. Can definitely cause issues, but this requires a larger rethinking potentially that should be in a proposal issue.
- [x] Figure out why ZIR indices on structs, unions, and enums seem to disappear (there's probably a good reason for this that I'm not aware of, or a bug)
- [x] Add parent_fix from CBE logic I stole back to remove unnecessary parens
- [x] Extract param names from functions
- [x] Better styling for output code (a little cramped? maybe group functions together?)
- [x] ~~Probably add a couple tests~~ Not blocking for merge, can be added later.

But it does work 100% better than the current emit-h! :P

### CODE REVIEW

**Comment by mlugg** [comment]
> Figure out why ZIR indices on structs, unions, and enums seem to disappear

If you're referring to the `zir_decl_index` field on `Decl`, this is by design -- the doc comment explains what's going on. Unfortunately, this code still overuses `Decl` -- this type currently has a lot of responsibilities, and I'm [actively working](https://github.com/ziglang/zig/pull/20299) on splitting it into some more atomic types. However, that need not block this PR -- from a quick glance, this work is a good baseline which I can build on as a part of my work on `Decl`.

**Review by mlugg — APPROVED** [review]
There are several TODOs you've noted, and a few more potential issues I've flagged up. However, this change overall seems pretty good. Since `emit-h` is completely broken today, I think this PR is reasonable to merge as-is to unblock at least some use cases (such as your own!).

Upon merge, I will open a follow-up issue detailing the remaining issues with `emit-h`. I will likely end up solving some of them in my upcoming `Decl` refactor, which is another reason I think this should be merged as-is.

**Comment by jacobly0** [comment]
Suggested future work:
1. I'm pretty sure emit-h should not be using `zig.h` (and as has been pointed out, the type emulation code performed by `zig.h`  is not appropriate for emit-h at all), the amount of support code actually required by emit-h is very minimal and the parts that are actually used can be conditionally written to the top of the file.
2. There's no chance that type decls are being emitted in the correct dependency order without a type pool.  The emit-h-specific type transformations should be a separate mode when adding to the pool, not special-cased here.
3. Why is this using a generic sort to perform a bucket sort, preferably just keep them in separate data structures per section to begin with.

**Comment by SuperAuguste** [comment]
Thanks for the feedback folks. I will try to address everything today.

**Comment by andrewrk** [comment]
draft status, no updates in 30+ days

**Comment by squeek502** [comment]
@wainejr

> SuperAuguste marked this pull request as draft 4 months ago

See [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests) if you're not already familiar with what a draft pull request is.

**Pull request closed without merging**
